library/std/src/sys/itron/thread.rs

```diff
@@ -17,12 +17,18 @@ use crate::{
 };
 
 pub struct Thread {
-    inner: ManuallyDrop<Box<ThreadInner>>,
+    p_inner: *mut ThreadInner,
 
     /// The ID of the underlying task.
     task: abi::ID,
 }
 
+// Safety: There's nothing in `Thread` that ties it to the original creator. It
+//         can be dropped by any threads.
+unsafe impl Send for Thread {}
+// Safety: `Thread` provides no methods that take `&self`.
+unsafe impl Sync for Thread {}
```

Comment:
It's not intuitive to me that implementing a function or not that takes `&self` affects safety for this case, as borrowing is a primitive notion in Rust, so anyone can hypothetically take `&Thread` from a given `Thread` and run off with it? Could you explain that one?

Reply:
This is the same reasoning as the one for [`Exclusive`](https://doc.rust-lang.org/nightly/core/sync/struct.Exclusive.html#): since you cannot do anything with the reference, the struct can be `Sync` without causing any UB.

I'm not sure this is the best reasoning here, but it is correct.

Reply:
Oh I see, so it's actually talking about `Thread` as a "smart pointer" (wrapper around the pointer). It interposes and blocks access to the *pointer inside*, and then never allows you to touch that inner pointer using a method that would take `&Thread`. Thus it's correct to `unsafe impl Sync for Thread`, since *the behavior of* the type can only be accessed via `&mut Thread` or `Thread`.

Yeah for some reason that didn't actually register immediately when I read it.